Add wait_policy option to EmrCreateJobFlowOperator.

adrian-adikteev opened the pull request:
<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---

Currently, when using `wait_for_complition=True` it will trigger the node to only wait until the cluster is successfully launched. 
I added the option `wait_policy` so that one can chose what waiter will be used. 

Possible values:
- `None`: No wait (default)
- `WaitPolicy.WAIT_FOR_COMPLETION`: Previous behaviour when `wait_for_completion` was `True`
- `WaitPolicy.WAIT_FOR_STEPS_COMPLETION`: New behaviour - wait for the cluster to terminate.

This is useful because it will eliminate the need to have an extra node in your DAG.

vincbeck reviewed (commented):
Could you cover this new use case in a unit test?

eladkal commented on providers/src/airflow/providers/amazon/aws/operators/emr.py:
```diff
@@ -666,6 +670,7 @@ def __init__(
         job_flow_overrides: str | dict[str, Any] | None = None,
         region_name: str | None = None,
         wait_for_completion: bool = False,
+        wait_for_steps_completion: bool = False,
```
I think both `wait_for_completion` and `wait_for_steps_completion` can be very confusing.
If we are to support the new functionality I suggest to introduce a new parameter `wait_policy` which will accept value of `wait_for_steps_completion`, `wait_for_steps_completed_and_machine_terminated` similar to what we did with Kubernetes in https://github.com/apache/airflow/pull/30718 when we had more than just a bool option for delete pod.

vincbeck replied:
I like that idea!

adrianlzr replied:
@eladkal I prefer your idea too. I've added a version of what you said, but it might not be what you had in mind.

When you say `wait_for_steps_completed_and_machine_terminated` what is the waiter name you think should map to it?

eladkal commented on providers/src/airflow/providers/amazon/aws/operators/emr.py:
```diff
@@ -638,6 +642,10 @@ class EmrCreateJobFlowOperator(BaseOperator):
     :param region_name: Region named passed to EmrHook
     :param wait_for_completion: Whether to finish task immediately after creation (False) or wait for jobflow
         completion (True)
```
Given the new functionality we will need to deprecate `wait_for_completion` and add backward compatibility to the new functionality. You can see similar example in https://github.com/apache/airflow/pull/30718

adrianlzr replied:
Alright. So that's what you meant.

adrianlzr replied:
Done

eladkal commented on providers/src/airflow/providers/amazon/aws/utils/waiter.py:
```diff
@@ -83,3 +84,16 @@ def get_state(response, keys) -> str:
+
+class WaitPolicy(str, Enum):
+    # Wait for the cluster to be up.
+    DEFAULT = "default"
```
Lets not say default but be explicit about what it is.
I think we have 
`wait_for_completion` - equivalent to what we have now
`wait_for_completion_with_machine_terminated` - the new functionality you wanted

You can suggest better names

adrianlzr replied:
That's good.

adrianlzr replied:
Done

eladkal commented on providers/src/airflow/providers/amazon/aws/operators/emr.py:
```diff
@@ -636,8 +641,14 @@ class EmrCreateJobFlowOperator(BaseOperator):
+        - wait for the jobflow completion (WaitPolicy.WAIT_FOR_COMPLETION)
+        - wait for the jobflow completion and cluster to terminate (WaitPolicy.WAIT_FOR_STEPS_COMPLETION)
+        (default: None)
     :param waiter_max_attempts: Maximum number of tries before failing.
```
Can you add the explnation about the parameter also to the emr docs?
Also lets verify if we have example dags using the old parameter and switch them to the new one

vincbeck replied:
> Also lets verify if we have example dags using the old parameter and switch them to the new one

I did check and we do not :)

adrianlzr replied:
No examples using the old parameter. 

I've updated the doc to describe `wait_policy`

adrian-adikteev committed:
Add wait_policy option to EmrCreateJobFlowOperator.

Possible values:

- None: No wait (default)
- WaitPolicy.WAIT_FOR_COMPLETION: Previous behaviour when wait_for_completion was True
- WaitPolicy.WAIT_FOR_STEPS_COMPLETION: New behaviour - wait for the cluster to terminate.

vincbeck commented on docs/apache-airflow-providers-amazon/operators/emr/emr.rst:
```diff
@@ -47,6 +47,15 @@ Create an EMR job flow
+It is possible to modify this behaviour by using a different ``wait_policy``. Available options are:
+
+- ``WaitPolicy.WAIT_FOR_COMPLETION`` - DAG Task node waits for the cluster to be Running
+- ``WaitPolicy.WAIT_FOR_STEPS_COMPLETION`` - DAG Task node waits for the cluster to Terminate
```
nits

```suggestion
- ``WaitPolicy.WAIT_FOR_COMPLETION`` - DAG Task node waits for the cluster to be running
- ``WaitPolicy.WAIT_FOR_STEPS_COMPLETION`` - DAG Task node waits for the cluster to terminate
```

adrianlzr replied:
Done

vincbeck commented on newsfragments/44055.feature.rst:
```diff
@@ -0,0 +1 @@
+New argument ``wait_policy`` to control waiting behaviour when using ``EmrCreateJobFlowOperator``.
```
You do not need this file, it is not a breaking change and more importantly, it is a provider change, not an Airflow change. Thus, you can just remove this file :)

adrianlzr replied:
Alright. Done

vincbeck reviewed (changes requested):
Overall looks good! Just one nit and you do not a newsfragment for this change

vincbeck reviewed (approved):
Perfect!
